### ISSUE TITLE
Crafting a self-contained deployment 

### ISSUE THREAD

**Issue by fearthecowboy** [dotnet/core-setup#1519]
I'm in the process of crafting our deployment for our project, and I'd like to be able to split the framework from the application binaries.

I have reviewed https://docs.microsoft.com/en-us/dotnet/articles/core/preview3/deploying/#self-contained-deployments-scd alot.

I would _**not**_ like to use the system-level shared framework - I do not want to require root/admin to install.

I have the necessity of having multiple versions of our project to be installed side-by-side, but I certainly don't want copies of the same framework over and over (not to mention that I really don't want to ship the whole framework in every single build (nightlys, etc) 

Ideally, I'd like to do something like:

``` text
/folder/to/autorest
    + framework
        +1.0.4
        +1.1
    + app
        + 1.0
        + 1.0.2-nightly
        + 1.1
   
```

Where I could prepare the x-copy-able  `framework` separately from the `app` and ship them in individual zip files. Then when I install a new version of the app SxS, I don't have to drag a new framework along.

**My question is: ** what magic combination of `environment variables` | `command-line-switches` | `json-file-voodoo` do I need to make it so I can run the app from one folder and reference a private framework from another?

**Comment by Petermarcu** [comment]
Just make sure the first "dotnet.exe" on the path is the one in the location of the private framework and all dotnet commands will run in the context of that dotnet.exe.

Get the tar.gz for the runtime from here: https://www.microsoft.com/net/download/linux.
Set the PATH properly as I said above of just call the right dotnet version and then pass it the app you want it to run on that runtime.

**Comment by fearthecowboy** [comment]
Looks pretty good....

Quick question -- I notice it's picking up assemblies for libraries that I don't have in the output folder from the nuget package cache. 

I see that I can use `--additionalprobingpath ../shared/packages` and tell it to find packages from there. Is there a way to tell it *not* to search the global package cache?  

I'm assuming that explicitly stated `--additionalprobingpath` is searched first, so maybe it's not that big of deal.

**Comment by fearthecowboy** [comment]
And can I rely on those fwlinks to point to the specific version of those files?

**Comment by Petermarcu** [comment]
It's picking things up from the .nuget cache because you didn't publish. Build doesn't get you a final shippable thing, you have to publish. Once you do that, you will get your dependencies with you.

Do you mean, can you put the links in your scripts to pull them down? I would put them somewhere you own.

**Comment by Petermarcu** [comment]
I can also get you some links you can rely on if you want :)

**Comment by fearthecowboy** [comment]
DERP. Ah, yeah, I was in the `output` dir, not the `publish` dir under that. Nice. 

Yeah, I'd love some links that will be immutable; we're doing this under the covers in a VSCode extension, and I certainly don't want it to die obscurely if the framework link changes :D  

Hmm. I'm nearly to the point of success 😄

**Comment by Petermarcu** [comment]
Which links do you need? Windows and Ubuntu 16.10? For .NET Core 1.0.4 and 1.1.0?

**Comment by fearthecowboy** [comment]
Realistically, I need them all. 

I've swiped the platform detection and installer logic from the omnisharp-csharp extension, and our tools should run everywhere vscode does (windows, Debian, Ubuntu, Redhat, fedora, opensuse ,centos, macos).

Yeah, 1.0.4 and 1.1.0 would be great.

**Comment by Petermarcu** [comment]
1.0.4 hasn't shipped yet. I assume you are just grabbing that from our master branch? Do want to be on 1.0.3? 

There is a good script for getting the right thing on the distro that @piotrpMSFT is going to help me understand how to use better. Its here: 

https://github.com/dotnet/cli/blob/rel/1.0.0/scripts/obtain/dotnet-install.sh

It helps resolve what I'm about to describe...

These are example links for Debian. This is what we do to create our Docker Images and we keep them stable:

For 1.0.X:
https://dotnetcli.blob.core.windows.net/dotnet/preview/Binaries/1.0.3/dotnet-debian-x64.1.0.3.tar.gz

For 1.1.X:
https://dotnetcli.blob.core.windows.net/dotnet/release/1.1.0/Binaries/1.1.0/dotnet-debian-x64.1.1.0.tar.gz

From there, you just need to replace the debian-x64 part with each of the distros you care about. You can look at the links here to pull the OS string out for each one.

https://github.com/dotnet/core-setup 

I'm hoping @piotrpMSFT 's solution will help make this easier.

**Comment by fearthecowboy** [comment]
Awesome thanks!

**Comment by blackdwarf** [comment]
The script should be a preferred way to get the CLI and/or the runtime in these kinds of scenarios, mostly because it also downloads from the CDN. You can actually see the specc for the overall CLI acquisition scenarios on https://github.com/dotnet/cli/blob/rel/1.0.0/Documentation/cli-installation-scenarios.md and it covers the installation script. We also have docs for it on https://docs.microsoft.com/en-us/dotnet/articles/core/preview3/tools/dotnet-install-script.

**Comment by fearthecowboy** [comment]
The script isn't really a fantastic way to automate the installation. Further to that, it doesn't solve any of the problems we actually have...

I was under the (apparently mistaken) impression that there was the notion of a "self-contained deployment" -- which there is, only if you have already installed the pre-requisites mentioned at https://github.com/dotnet/core/blob/master/Documentation/prereqs.md   (WHICH-- only lists **Ubuntu** and **Centos** -- what about other Linux variants? )

This really kinda throws our plans into the toilet.  We were aiming to install the dotnet framework into a folder under the user's home folder, and having the whole thing self-contained. 

Having to install some things as root kinda puts a crimp in that, given that for some scenarios we're doing this in a language service inside of Visual Studio Code.  

In Windows, we don't have to elevate to install our framework to install the user-local framework.


Plus, given the dependencies: 

```
# ubuntu
libunwind8
libunwind8-dev
gettext
libicu-dev
liblttng-ust-dev
libcurl4-openssl-dev
libssl-dev
uuid-dev
unzip
```

Which of these are needed for running apps as opposed to what's needed for the SDK-- we're not installing the SDK, just the framework. Do we still need all those `-dev ` libraries?

```
# centos
deltarpm
epel-release
unzip
libunwind
gettext
libcurl-devel
openssl-devel
zlib
libicu-devel
```
And for `centos` -- do we really need deltarpm for just running dotnet apps?!

**Comment by fearthecowboy** [comment]
@Petermarcu Is it physically possible for me to gather up all the system-level dependent bits and merge them into the framework so that I don't have to do system-level installs of those things?

**Comment by Petermarcu** [comment]
Yes, @janvorli is working on making sure you can do that. I think there were a couple bugs he had to fix to make sure we load them but would love to have you try it out :)

Note, it's only going to work for .NET Core 2.0+.

**Comment by janvorli** [comment]
@fearthecowboy regarding the list of dependencies, you are right. We need to fix our documentation to describe separately the runtime dependencies and the build time (meaning coreclr / corefx build) dependencies. Basically, you don't need any of the dev packages for runtime, just their corresponding non-dev variants.

**Comment by fearthecowboy** [comment]
@Petermarcu  At this point, I'm more concerned with making sure it all installs smoothly, rather than which version of .NET Core I'm attached to. 😃 

I'll take whatever @janvorli has, right away!

**Comment by Petermarcu** [comment]
This is where I usually point people :) https://github.com/dotnet/dotnet-docker/blob/master/1.1/debian/runtime-deps/Dockerfile

Thats what is actually needed at runtime for debian.

@janvorli, we should see whether @fearthecowboy can try the work you are doing to enable people to copy the linux binaries into their app and point to them to alleviate the deployment pain.
